middle_end/flambda/types/env/typing_env_level.rec.ml

```diff
@@ -277,41 +277,48 @@ let concat (t1 : t) (t2 : t) =
     symbol_projections;
   }
 
-let join_types ~env_at_fork envs_with_levels ~extra_lifted_consts_in_use_envs =
+let join_types ~env_at_fork envs_with_levels =
   (* Add all the variables defined by the branches as existentials to the
      [env_at_fork].
      Any such variable will be given type [Unknown] on a branch where it
      was not originally present.
      Iterating on [level.binding_times] instead of [level.defined_vars] ensures
-     consistency of binding time order in the branches and the result. *)
+     consistency of binding time order in the branches and the result.
+     In addition, this also aggregates the code age relations of the branches.
+  *)
   let env_at_fork =
-    List.fold_left (fun env_at_fork (_, _, _, level) ->
-        Binding_time.Map.fold (fun _ vars env ->
-            Variable.Set.fold (fun var env ->
-                if Typing_env.mem env (Name.var var) then env
-                else
-                  let kind = Variable.Map.find var level.defined_vars in
-                  Typing_env.add_definition env
-                    (Name_in_binding_pos.var
-                       (Var_in_binding_pos.create var Name_mode.in_types))
-                    kind)
-              vars
-              env)
-          level.binding_times
-          env_at_fork)
+    List.fold_left (fun env_at_fork (env_at_use, _, _, level) ->
+        let env_with_variables =
+          Binding_time.Map.fold (fun _ vars env ->
+              Variable.Set.fold (fun var env ->
+                  if Typing_env.mem env (Name.var var) then env
+                  else
+                    let kind = Variable.Map.find var level.defined_vars in
+                    Typing_env.add_definition env
+                      (Name_in_binding_pos.var
+                         (Var_in_binding_pos.create var Name_mode.in_types))
+                      kind)
+                vars
+                env)
+            level.binding_times
+            env_at_fork
+        in
+        let code_age_relation =
+          Code_age_relation.union (Typing_env.code_age_relation env_at_fork)
+            (Typing_env.code_age_relation env_at_use)
+        in
+        Typing_env.with_code_age_relation env_with_variables code_age_relation)
       env_at_fork
       envs_with_levels
   in
   (* Now fold over the levels doing the actual join operation on equations. *)
   ListLabels.fold_left envs_with_levels
-    ~init:(env_at_fork, Name.Map.empty, true)
-    ~f:(fun (join_env, joined_types, is_first_join) (env_at_use, _, _, t) ->
-      let join_env =
-        Code_age_relation.union (Typing_env.code_age_relation join_env)
-          (Typing_env.code_age_relation env_at_use)
-        |> Typing_env.with_code_age_relation join_env
+    ~init:(Name.Map.empty, true)
+    ~f:(fun (joined_types, is_first_join) (env_at_use, _, _, t) ->
+      let left_env =
```

Comment:
We might want to add a comment mentionning that there might be some cost here (to look if there is something too expensive hapennening aroung join)